util/topsql/reporter/reporter.go

```diff
@@ -259,45 +254,50 @@ func (d *dataPoints) merge(other *dataPoints) {
 		return
 	}
 	if other.isInvalid() {
-		logutil.BgLogger().Warn("[top-sql] data points is invalid, it should never happen", zap.Reflect("self", *d), zap.Reflect("other", *other))
+		logutil.BgLogger().Warn("[top-sql] data points is invalid, it should never happen", zap.Any("self", d), zap.Any("other", other))
 		return
 	}
-	if !sort.IsSorted(other) {
-		// This situation appears only when time jump backward.
-		// If so, sort the data points by timestamp.
-		sort.Sort(other)
-	}
+	// Sort the dataPoints by timestamp to fix the affect of time jump backward.
+	sort.Sort(d)
+	sort.Sort(other)
 	if len(d.TimestampList) == 0 {
 		d.TimestampList = other.TimestampList
 		d.CPUTimeMsList = other.CPUTimeMsList
 		d.CPUTimeMsTotal = other.CPUTimeMsTotal
 		return
 	}
-	d.CPUTimeMsTotal += other.CPUTimeMsTotal
-	findIdx := 0
-	for idx, ts := range other.TimestampList {
-		cpuTimeMs := other.CPUTimeMsList[idx]
-		for ; findIdx < len(d.TimestampList); findIdx++ {
-			findTs := d.TimestampList[findIdx]
-			if ts == findTs {
-				d.CPUTimeMsList[findIdx] += cpuTimeMs
-				break
-			}
-			if ts < findTs {
-				d.TimestampList = append(d.TimestampList, 0)
-				copy(d.TimestampList[findIdx+1:], d.TimestampList[findIdx:])
-				d.TimestampList[findIdx] = ts
-				d.CPUTimeMsList = append(d.CPUTimeMsList, 0)
-				copy(d.CPUTimeMsList[findIdx+1:], d.CPUTimeMsList[findIdx:])
-				d.CPUTimeMsList[findIdx] = cpuTimeMs
-				break
-			}
-		}
-		if ts > d.TimestampList[len(d.TimestampList)-1] {
-			d.TimestampList = append(d.TimestampList, ts)
-			d.CPUTimeMsList = append(d.CPUTimeMsList, cpuTimeMs)
+	length := mathutil.Max(len(d.TimestampList), len(other.TimestampList))
```

Comment:
I think you should set the *preserved* length to `len(d.ts) + len(other.ts)` instead of `max(len(d.ts), len(other.ts))` since it is the maximum length possible.

Consider this case:

```go
d.ts = [1, 2, 3]
other.ts = [4, 5, 6]
```

After merging the length should be 6. So that if we preserve 6 elements then no re-allocation will happen.